Вариант 2.

MikhailDergachev opened the pull request:
Написать программу, в которой описана иерархия классов: человек («дошкольник», «школьник», «студент», «работающий»). Базовый класс должен иметь поля для хранения ФИО, возраста, пола, а также методы получения среднего дохода в денежном эквиваленте. Продемонстрировать работу всех методов классов, предоставив пользователю выбор типа объекта для демонстрации.
Создал иерархию классов.
«дошкольник»
    ФИО
«школьник»
    возраста
«студент»
    пол
«работающий»
    методы получения среднего дохода в денежном эквиваленте(Средний дн. заработок = зарплата / 29,3 х 12)

MikhailDergachev committed:
Вариант 2.

Написал Api, реализацию и main.
Создал иерархию классов.
«дошкольник»
    ФИО
«школьник»
    возраста
«студент»
    пол
«работающий»
    методы получения среднего дохода в денежном эквиваленте(Средний дн. заработок = зарплата / 29,3 х 12)

MarinaPaley commented on preschooler.cpp:
```diff
@@ -0,0 +1,54 @@
+#include <iostream>
+#include "preschooler.h"
+
+preschooler::preschooler(const std::string name, const std::string surname, const std::string second_name)
```
`std:;string&`

MarinaPaley commented on preschooler.cpp:
```diff
@@ -0,0 +1,54 @@
+
+void preschooler::set_name(const std::string name)
+{
+    this->name = name;
```
а если `name == '""`?

MarinaPaley commented on preschooler.cpp:
```diff
@@ -0,0 +1,54 @@
+
+std::ostream& operator << (std::ostream& out, const preschooler& vector)
+{
+    return out << vector.to_string();
```
не то скопировали

MarinaPaley reviewed (changes requested):
Базовая сущность - человек с полями ФИО и возраст, остальные от него наследуются

MikhailDergachev committed:
Delete main.cpp

MikhailDergachev committed:
Delete preschooler.cpp

MikhailDergachev committed:
Delete preschooler.h

MikhailDergachev committed:
Delete schooler.cpp

MikhailDergachev committed:
Delete schooler.h

MikhailDergachev committed:
Delete student.cpp

MikhailDergachev committed:
Delete student.h

MikhailDergachev committed:
Delete worker.cpp

MikhailDergachev committed:
Delete worker.h

MikhailDergachev committed:
Переписал с нуля весь код.

Учитывая ошибки предыдущего кода, написал зависимость всех классов от human.

MikhailDergachev committed:
Исправил cpp файл

MikhailDergachev committed:
Исправил хедер

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+#include <iostream>
+#include "human.h"
+
+human::human(const std::string name, const std::string surname, const std::string second_name, const double age, const bool gender,const double salary)
```
Объекты типа `string` необходимо передавать по ссылке

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+    if (name == "") {
+        throw;
+    }
+    return this->name;
```
зачем? если это сеттер?

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+    if (surname == "") {
+        throw;
+    }
+    return this->surname;
```
?

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+    if (second_name == "") {
+        throw;
+    }
+    return this->second_name;
```
?

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+    if (age <= 0) {
+        throw;
+    }
+    return this->age;
```
?

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+    } else {
+        _gender = "Женщина";
+    }
+};
```
ООП в действии! Вы его не понимаете. Зачем передавать в метод СВОЕ же поле?

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+
+void set_salary(const double salary)
+{
+    return this->salary;
```
неправильно

MarinaPaley commented on human.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+double average_income(const double salary) {
+    return salary/29.3*12;
```
что это такое?
Не надо использовать _&laquo;магические константы&raquo;_ в коде. Это не только признак дурновкусия, но и источник большого числа потенциальных ошибок. Привожу некоторые источники, потребные для Вашего самообразования:
- [ru.wikipedia.org](https://ru.wikipedia.org/wiki/Магическое_число_(программирование))
- [refactoring.guru](https://refactoring.guru/ru/replace-magic-number-with-symbolic-constant)
- [pvs-studio.com](https://pvs-studio.com/ru/blog/posts/cpp/0023/)
- [habr.com](https://habr.com/ru/post/328244/)

MarinaPaley reviewed (changes requested):
У каждого класса должен быть свой файл (два!)
ООП не понимаете

MikhailDergachev committed:
Delete human.cpp

MikhailDergachev committed:
Delete human.h

MikhailDergachev committed:
Delete main.cpp

MikhailDergachev committed:
Третий вариант написания программы:)

Ознакомившись с разными материалами по ООП, поправил программу, в соответствии с вашими требованиями. Принял решение разделить программу на разные файлы, чтобы можно было четче видеть структуру программы. Все основные методы описаны в human.cpp . Остальные cpp файлы по условию задания не содержат методов, но я оставил их на всякий случай и для  четкой связи между хедером и cpp файлом по типу : если есть heder, значит должен быть cpp файл. Программу проверил, она работает полностью исправно.

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
+public:
     /**
-     * @brief Имя
+     * @brief ����������� �� ���������
```
кодировка не UTF-8

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
      */
-    human::human();
-
+    const char* get_error_message() const;
```
не нужно использовать стандарт С++ ниже, чем С++11
используйте тип данных `std::string`, а не небезопасный тип данных `char*`

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
+     * @param second_name �������� ��������
      */
-    std::string get_gender() const;
+    explicit human(const std::string& name, const std::string& surname,
```
зачем тип данных `double` Вы передаете по ссылке?

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
-
-class preschooler: public human{
-    public:
+    ~human() = default;
```
[Всегда определяйте деструктор `virtual` для полиморфного разрушения объекта](https://stackoverflow.com/questions/461203/when-to-use-virtual-destructors)

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
+     * @param salary ����� �����
      */
-    ~worker() = default;
+    void set_salary(const double& salary);
```
зачем передавать `double` по ссылке, Вы данный параметр собираетесь изменять?

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
-     * @return переназначенное name
+     * @brief ��������
      */
+    double salary_;
```
это не ответственность `human`, данное поле не нужно в базовом классе, а нужно в соответсвующем классе наследнике

MarinaPaley commented on human.h:
```diff
@@ -2,249 +2,181 @@
      */
-    worker(const std::string& name, const std::string& surname, const std::string& second_name, const double age,const std::string& gender,const double salary)
-    : human(name,surname,second_name,age,gender,salary){};
+    double average_income() const;
```
метод нужен в том классе, в котором будет поле salary`

MarinaPaley reviewed (changes requested):
неправильное наследование

MikhailDergachev committed:
Delete human.cpp

MikhailDergachev committed:
Delete human.h

MikhailDergachev committed:
Delete main.cpp

MikhailDergachev committed:
Delete preschooler.cpp

MikhailDergachev committed:
Delete preschooler.h

MikhailDergachev committed:
Delete schooler.cpp

MikhailDergachev committed:
Delete schooler.h

MikhailDergachev committed:
Delete student.cpp

MikhailDergachev committed:
Delete student.h

MikhailDergachev committed:
Delete worker.cpp

MikhailDergachev committed:
Delete worker.h

MikhailDergachev committed:
Подправил недочеты.

Исправил ошибки, основываясь на комментариях
